### ISSUE TITLE
sh: husky: command not found

### ISSUE THREAD

**Issue by GaelFerrand** [typicode/husky#945]
**Context**

I've setup a node project with husky but when my collegue tries to run `npm install` on his Mac he gets the following error : 

```
noa-be@1.0.0 prepare
husky install

sh: husky: command not found
npm ERR! code 127
npm ERR! path /Users/X/Desktop/Workspace/project
npm ERR! command failed
npm ERR! command sh -c husky install

npm ERR! A complete log of this run can be found in:
npm ERR!     /Users/X/.npm/_logs/2021-04-12T13_07_25_842Z-debug.log
```

These are the relevant package.json parts: 

```json
{
    "scripts": {
        "prepare": "husky install"
    },
    "devDependencies": {
        "husky": "^5.2.0",
    }
}
```

I thought this would be enough for husky to be installed when running `npm install`, but it's not. What am I missing?

Thank you!

**Comment by flo-sch** [comment]
Seems odd, is husky successfully being installed in the node_modules folder?

(Is there nothing in his environment preventing devDependencies to be installed?)

**Comment by OthrysDev** [comment]
Husky is in the node_modules!

**Comment by GaelFerrand** [comment]
Husky is in the node_modules!

**Comment by DoctorDerek** [comment]
Is he running the command as simply `prepare`? It's a bit hard to tell from the code you shared.

When you do `npm run prepare` then it uses the local node_modules folder on the path, in the same that `npx husky install` would work.

**Comment by DoctorDerek** [comment]
I saw a similar error when I tried to commit a change without having run npm install in the current working directory, but it wasn't a husky issue -- it was husky itself flagging prettier as missing.

**Comment by drazisil** [comment]
```
/home/drazisil/.nvm/versions/node/v15.14.0/bin/npm: npm: line 2: syntax error near unexpected token `'../lib/cli.js''
/home/drazisil/.nvm/versions/node/v15.14.0/bin/npm: npm: line 2: `require('../lib/cli.js')(process)'
```

The error alone doesn't tell me anything. I suspect this is related to setting `type: module` in my package.json.

**Comment by rdmeal** [comment]
I had a very similar error message to this when running `npm install` inside a Docker container and it was down to the node environment being set to production - see #914 for more information and a possible solution.

**Comment by alexbu92** [comment]
Having this problem also, no solution found. husky is in node_modules.

**Comment by Spartakyste** [comment]
For me, the problem was husky file not being executables by default. Husky was properly installed.

I'm on Mac (not sure if the same happen on other OS), and last git version helped me to identify the problem. (V 2.24.1 as time of writing this).

To give executable right on your file : `chmod +x PATH_TO_HUSKY_FILE`

You'll need to execute it for every hook you have

**Comment by vintprox** [comment]
Everything seems alright on local machine, but when I deploy on Heroku, I get this:
```
...
remote: -----> Installing dependencies
remote:        Installing node modules
remote:
remote:        > drcm-bot@0.1.0 prepare /tmp/build_b8e6fb4d
remote:        > husky install
remote:
remote:        sh: 1: husky: not found
remote:        npm ERR! code ELIFECYCLE
remote:        npm ERR! syscall spawn
remote:        npm ERR! file sh
remote:        npm ERR! errno ENOENT
remote:        npm ERR! drcm-bot@0.1.0 prepare: `husky install`
remote:        npm ERR! spawn ENOENT
...
```

Cannot find a solution. Help.

**Comment by ctsstc** [comment]
I'm noticing this same issue, it seems that having it as a devDependency doesn't work if you're only installing production only packages. This seems to have been working fine until I updated npm. Ideally you shouldn't need Husky for a production build though, but does `prepare` run always? I tried looking up more about it in the npm documentation and it says it runs "locally" but doesn't explain what changes if it's "local" or not.

https://docs.npmjs.com/cli/v8/using-npm/scripts#:~:text=Runs%20on%20local%20npm%20install%20without%20any%20arguments

I can replicate this locally by doing the following which is equivalent to a production build:
```
> npm install --omit=dev

>project@1.1.1 prepare
> husky install

sh: husky: command not found
npm ERR! code 127
npm ERR! path /Users/coder/projects/project
npm ERR! command failed
npm ERR! command sh -c husky install

npm ERR! A complete log of this run can be found in:
npm ERR!     /Users/coder/.npm/_logs/2021-11-30T23_50_31_403Z-debug.log
```

**Comment by mindctrl** [comment]
I ran into this error. Turns out something on my laptop was setting `NODE_ENV="production"`, and husky wasn't being installed because it's a dev dependency. Setting `NODE_ENV` to `development` fixed the problem.

**Comment by ctsstc** [comment]
I seem to have no problems on a new project with:
- node: `v16.13.2`
- npm: `8.1.2`
- husky: `7.0.4`

On the other project when I originally commented here I ended up doing: 
```json
"prepare": "if [[ $NODE_ENVIRONMENT != \"production\" ]]; then husky install; fi"
```

**Comment by raphaelsetin** [comment]
@ctsstc That didn't work for me.

The following worked (and is cross platform too):
```json
"prepare": "node -e \"if (process.env.NODE_ENV !== 'production'){process.exit(1)} \" || husky install",
```

**Comment by NachtRitter** [comment]
If you use command like `npm ci --omit=dev` you can access those property via `$npm_config_omit`.
According to value of `$npm_config_omit` you can decide to run husky install command or not.

```
{
  "postinstall": "if [ \"$npm_config_omit\" = \"dev\" ]; then echo \"Skip husky installation\"; else husky install; fi"
}
```

**Comment by razvanphp** [comment]
Another way:
```
  "scripts": {
    "prepare": "husky install || true",
```

**Comment by olange** [comment]
I ran into the same error on Vercel, and found this issue. In case anyone comes here for the same reason, although not specifically linked to the issue above, this might help:

My issue was that I am using NPM workspaces within a mono repo, and I set the _root directory_ of the Vercel project to `docs/` subfolder (one of the workspaces) — which has its own `package.json` with *no Husky* in its dev dependencies.

Resetting the _root directory_ of the Vercel project to the _root_ of the monorepo solved the issue: `npm install` running from the root has `husky` in its dev deps and the `"prepare": "husky"` scritpt runs smoothly.
